internal/plugin/connectors/http/generic/oauthprotocol/protocol.go

```diff
@@ -0,0 +1,214 @@
+	r.Body = copyBody
+
+	// these keys are already URL safe
+	paramMap[oauth_consumer_key] = []string{oauth1.ConsumerKey}
+	paramMap[oauth_nonce] = []string{oauth1.Nonce}
+	paramMap[oauth_signature_method] = []string{oauth1.SignatureMethod}
+	paramMap[oauth_timestamp] = []string{oauth1.TimeStamp}
+	paramMap[oauth_token] = []string{oauth1.Token}
+	paramMap[oauth_version] = []string{oauth1.Version}
+
+	// make map of keys and sort to append lexicographically later
+	keys := make([]string, 0, len(paramMap))
+	for k := range paramMap {
+		keys = append(keys, k)
+	}
+	sort.Strings(keys)
+
+	// create string
+	for indK, key := range keys {
+		// URL encode all values in all keys (most should be 1, but this double for loop is still gross)
+		for ind, value := range paramMap[key] {
+			encodedValue := strings.ReplaceAll(url.QueryEscape(value), "+", "%20")
+			paramMap[key][ind] = encodedValue
+		}
+
+		// sort the array of URL encoded values in each key
+		sort.Strings(paramMap[key])
+
+		for indV, value := range paramMap[key] {
+			// Do not add '&' on first value
+			if indK != 0 || indV > 0 {
+				parameterString.WriteString("&")
+			}
+			keyValuePair := fmt.Sprintf("%s=%s", key, value)
+			parameterString.WriteString(keyValuePair)
+		}
+	}
+
+	return parameterString.String()
+}
+
+func generateBaseString(r *gohttp.Request, paramString string) string {
+	urlParams := []string{
+		r.Method,
+		encodeURI(r.URL.Scheme + "://" + r.URL.Host + r.URL.Path),
+	}
+	urlParamsString := strings.Join(urlParams, "&")
+	encodedParamString := encodeURI(paramString)
+	return fmt.Sprintf("%s&%s", urlParamsString, encodedParamString)
+}
+
+func generateSigningKey(oauth1 OAuth1) string {
+	return encodeURI(oauth1.ConsumerSecret) + "&" + encodeURI(oauth1.TokenSecret)
+}
+
+func hashSignature(signingKey string, baseString string) string {
+	h := hmac.New(sha1.New, []byte(signingKey))
+	h.Write([]byte(baseString))
+	return encodeURI(base64.StdEncoding.EncodeToString(h.Sum(nil)))
+}
+
+func constructOAuthString(oauth1 OAuth1) string {
+	return fmt.Sprintf("OAuth oauth_consumer_key=\"%s\", oauth_nonce=\"%s\", "+
+		"oauth_signature=\"%s\", oauth_signature_method=\"%s\", "+
+		"oauth_timestamp=\"%s\", oauth_token=\"%s\", oauth_version=\"%s\"",
+		oauth1.ConsumerKey, oauth1.Nonce, oauth1.Signature, oauth_signature_method_type,
+		oauth1.TimeStamp, oauth1.Token, oauth1.Version)
+}
+
+// CreateOAuth1Header calls a series of methods to
+// construct the OAuth1 'Authorization' Header String
+func CreateOAuth1Header(params map[string]string, r *gohttp.Request) (string, error) {
+
+	// check for required params
+	err := checkRequiredOAuthParams(params)
+	if err != nil {
+		return "", err
+	}
+
+	oauth1 := extractOAuthParams(params)
+
+	paramString := generateParameterString(oauth1, r)
+	baseString := generateBaseString(r, paramString)
+	signingKey := generateSigningKey(oauth1)
+
+	oauth1.Signature = hashSignature(signingKey, baseString)
+
+	oauthString := constructOAuthString(oauth1)
```

Comment:
Something tells me there's more ways this can error out outside of param extraction but for now not a big deal